internal/impl/ollama/chat_processor.go

```diff
@@ -287,14 +394,83 @@ func (o *ollamaCompletionProcessor) generateCompletion(ctx context.Context, syst
 	})
 	shouldStream := false
 	req.Stream = &shouldStream
-	var g string
-	err := o.client.Chat(ctx, &req, func(resp api.ChatResponse) error {
-		g = resp.Message.Content
-		return nil
-	})
-	return g, err
+	for _, t := range o.tools {
+		req.Tools = append(req.Tools, t.spec)
+	}
+	// Allow up to N iterations of calling tools
+	for range o.maxToolCalls + 1 {
+		var resp api.ChatResponse
+		o.logger.Tracef("making LLM chat request messages: %s", gabs.Wrap(req.Messages).EncodeJSON())
+		err := o.client.Chat(ctx, &req, func(r api.ChatResponse) error {
+			resp = r
+			return nil
+		})
+		if err != nil {
+			return "", err
+		}
+		if len(resp.Message.ToolCalls) == 0 {
+			return resp.Message.Content, nil
+		}
+		req.Messages = append(req.Messages, resp.Message)
+		for _, toolCall := range resp.Message.ToolCalls {
+			o.logger.Debugf("LLM requested tool %s with arguments: %s", toolCall.Function.Name, toolCall.Function.Arguments.String())
+			idx := slices.IndexFunc(o.tools, func(t tool) bool { return t.spec.Function.Name == toolCall.Function.Name })
+			if idx < 0 {
+				return "", fmt.Errorf("unknown tool call requested: %s", toolCall.Function.Name)
+			}
+			pipeline := o.tools[idx].pipeline
+			msg := service.NewMessage(nil)
+			msg.SetStructuredMut(map[string]any(toolCall.Function.Arguments))
+			output, err := service.ExecuteProcessors(ctx, pipeline, service.MessageBatch{msg})
+			if err != nil {
+				return "", fmt.Errorf("error calling tool %s: %w", toolCall.Function.Name, err)
+			}
+			resp, err := combineToSingleMessage(output)
+			if err != nil {
+				return "", fmt.Errorf("error processing pipeline %s output: %w", toolCall.Function.Name, err)
+			}
+			o.logger.Debugf("Tool %s response: %s", toolCall.Function.Name, resp)
+			req.Messages = append(req.Messages, api.Message{Role: "tool", Content: resp})
+		}
+	}
+	return "", fmt.Errorf("model did not finish after %d function calls", o.maxToolCalls)
```

Comment:
Would it perhaps be useful to return the partial result anyway along with the error?

Reply:
It'd just be a JSON object of what the next tool call is. I guess that could be helpful?

Reply:
Oh, right. Not that useful... People should fix their configs in that case.

Reply:
🤷